Source/WebCore/platform/mediastream/MediaStreamTrackPrivate.cpp

```diff
@@ -47,88 +48,353 @@
+        {
+            ASSERT(isMainThread());
+            return m_source->capabilities();
+        }
+
+        const RealtimeMediaSourceSettings& settings()
+        {
+            ASSERT(isMainThread());
+            return m_source->settings();
+        }
+
+        void start()
+        {
+            m_source->start();
+        }
+
+        void stop()
+        {
+            m_source->stop();
+        }
+
+        void requestToEnd()
+        {
+            m_shouldPreventSourceFromEnding = false;
+            m_source->requestToEnd(*this);
+        }
+
+        void setMuted(bool muted)
+        {
+            m_source->setMuted(muted);
+        }
+
+        void applyConstraints(const MediaConstraints& constraints, RealtimeMediaSource::ApplyConstraintsHandler&& completionHandler)
+        {
+            m_source->applyConstraints(constraints, WTFMove(completionHandler));
+        }
+
+        void postTask(Function<void()>&& task)
+        {
+            m_postTask(WTFMove(task));
+        }
+
+    private:
+        void sourceStarted() final
+        {
+            sendToMediaStreamTrackPrivate([] (auto& privateTrack) {
+                privateTrack.sourceStarted();
+            });
+        }
+
+        void sourceStopped() final
+        {
+            sendToMediaStreamTrackPrivate([captureDidFail = m_source->captureDidFail()] (auto& privateTrack) {
+                privateTrack.sourceStopped(captureDidFail);
+            });
+        }
+
+        void sourceMutedChanged() final
+        {
+            sendToMediaStreamTrackPrivate([muted = m_source->muted(), interrupted = m_source->interrupted()] (auto& privateTrack) {
+                privateTrack.sourceMutedChanged(interrupted, muted);
+            });
+        }
+
+        void sourceSettingsChanged() final
+        {
+            sendToMediaStreamTrackPrivate([settings = crossThreadCopy(m_source->settings()), capabilities = crossThreadCopy(m_source->capabilities())] (auto& privateTrack) mutable {
+                privateTrack.sourceSettingsChanged(WTFMove(settings), WTFMove(capabilities));
+            });
+        }
+
+        void sourceConfigurationChanged() final
+        {
+            sendToMediaStreamTrackPrivate([name = crossThreadCopy(m_source->name()), settings = crossThreadCopy(m_source->settings()), capabilities = crossThreadCopy(m_source->capabilities())] (auto& privateTrack) mutable {
+                privateTrack.sourceConfigurationChanged(WTFMove(name), WTFMove(settings), WTFMove(capabilities));
+            });
+        }
+
+        void hasStartedProducingData() final
+        {
+            sendToMediaStreamTrackPrivate([] (auto& privateTrack) {
+                privateTrack.hasStartedProducingData();
+            });
+        }
+
+        void sendToMediaStreamTrackPrivate(Function<void(MediaStreamTrackPrivate&)>&& task)
+        {
+            m_postTask([task = WTFMove(task), privateTrack = m_privateTrack] () mutable {
```

Comment:
Is there any reason to not null-check `m_privateTrack` and early return before invoking the task?

Reply:
We need to hop to the right thread to check m_privateTrack.